FileIO/MPI_IO/NodePartitionedMeshReader.cpp

```diff
@@ -0,0 +1,521 @@
+
+    MeshLib::NodePartitionedMesh* np_mesh = nullptr;
+    std::vector<MeshLib::Node*> mesh_nodes;
+    std::vector<unsigned long> glb_node_ids;
+    std::vector<MeshLib::Element*> mesh_elems;
+
+    for(int i = 0; i < _mpi_comm_size; i++)
+    {
+        if(_mpi_rank == 0)
+        {
+            INFO("-->Parallel reading the partitioned mesh: ");
+
+            // Read first part into _mesh_info which is equal with the binary
+            // structure.
+            for(unsigned long j = 0; j < 10; ++j)
+                is_cfg >> *(_mesh_info.data() + j);
+            // The last positon is the extra_flag.
+            is_cfg >> _mesh_info.extra_flag;
+            is_cfg >> std::ws;
+        }
+
+        MPI_Bcast(_mesh_info.data(), static_cast<int>(_mesh_info.size()),
+            MPI_LONG, 0, _mpi_comm);
+
+        //----------------------------------------------------------------------------------
+        // Read Nodes
+        if (!readCastNodesASCII(is_node, i, mesh_nodes, glb_node_ids))
+            break;
+
+        //----------------------------------------------------------------------------------
+        // Read elements
+        if (!readCastElemsASCII(is_elem, i,
+            _mesh_info.regular_elements + _mesh_info.offset[0],
+            false, mesh_nodes, mesh_elems))
+            break;
+
+        //-------------------------------------------------------------------------
+        // Ghost elements
+        if (!readCastElemsASCII(is_elem, i,
+            _mesh_info.ghost_elements + _mesh_info.offset[1],
+            true, mesh_nodes, mesh_elems))
+            break;
+
+        if(_mpi_rank == i)
+            np_mesh = newMesh(BaseLib::extractBaseName(file_name_base),
+                    mesh_nodes, glb_node_ids, mesh_elems);
+    }
+
+    if(_mpi_rank == 0)
+    {
+        is_cfg.close();
+        is_node.close();
+        is_elem.close();
+    }
+
+    MPI_Barrier(_mpi_comm);
+
+    return np_mesh;
+}
+
+MeshLib::NodePartitionedMesh*
+NodePartitionedMeshReader::newMesh(
+    std::string const& mesh_name,
+    std::vector<MeshLib::Node*> const& mesh_nodes,
+    std::vector<unsigned long> const& glb_node_ids,
+    std::vector<MeshLib::Element*> const& mesh_elems) const
+{
+    return new MeshLib::NodePartitionedMesh(
+        mesh_name + std::to_string(_mpi_comm_size),
+        mesh_nodes, glb_node_ids, mesh_elems,
+        _mesh_info.regular_elements,
+        _mesh_info.global_base_nodes,
+        _mesh_info.global_nodes,
+        _mesh_info.base_nodes,
+        _mesh_info.active_base_nodes,
+        _mesh_info.active_nodes);
+}
+
+void NodePartitionedMeshReader::readElementASCII(std::ifstream &ins,
+    std::vector<unsigned long>& elem_data, const bool ghost) const
+{
+    // Set number of elements.
+    unsigned long const ne =
+        ghost ? _mesh_info.ghost_elements : _mesh_info.regular_elements;
+    unsigned long counter = ne;
+    for(unsigned long j = 0; j < ne; j++)
+    {
+        elem_data[j] = counter;
```

Comment:
@wenqing Thanks for the explanation! Now I understand. Before it was not clear to me that the entries [0,ne-1[ are used as offsets. It would be very good to add this to the documentation of the method setElements in the header file and maybe mention here that the format elem_data is described there.